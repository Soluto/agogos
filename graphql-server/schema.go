package main

import (
	graphql "github.com/graphql-go/graphql"
	"github.com/vektah/gqlparser"
	"github.com/vektah/gqlparser/ast"

	"context"
	"fmt"
	"google.golang.org/grpc"
	"graphql-gateway/generated"
	"graphql-gateway/utils"
	"io"
)

const (
	address = "graphql-gateway.schema-registry:81"
)

type schemaResult struct {
	schema *graphql.Schema
	err    error
}

func subscribeToSchema(schemas chan schemaResult) (err error) {
	defer utils.Recovery(&err)

	conn, err := grpc.Dial(address, grpc.WithInsecure())
	if err != nil {
		fmt.Println("error initiating GRPC channel")
		return err
	}
	defer conn.Close()

	gqlSchemaClient := gqlschema.NewGqlSchemaClient(conn)

	stream, err := gqlSchemaClient.Subscribe(context.Background(), &gqlschema.GqlSchemaSubscribeParams{})
	if err != nil {
		fmt.Println("error subscribing to schema-registry")
		return err
	}

	for {
		gqlSchemaMessage, err := stream.Recv()

		if err == io.EOF {
			fmt.Println("got EOF")
			break
		}
		if err != nil {
			fmt.Println("error receiving message")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		astSchema, err := parseSdl(source{
			name: "schema regisrty sdl",
			sdl:  gqlSchemaMessage.Schema,
		})
		if err != nil {
			fmt.Println("error parsing SDL")
			fmt.Println("err", err)
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		schema, err := ConvertSchema(astSchema)
		if err != nil {
			fmt.Println("error converting schema")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		schemas <- schemaResult{
			schema: schema,
			err:    nil,
		}
	}
	return
}

type source struct {
	name string
	sdl  string
}

func parseSdl(s source) (*ast.Schema, error) {
	schema, err := gqlparser.LoadSchema(&ast.Source{
		Name:  s.name,
		Input: s.sdl,
	})

	if err != nil {
		return nil, err
	}

	return schema, nil
}
