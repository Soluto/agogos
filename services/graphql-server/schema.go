package main

import (
	graphql "github.com/graphql-go/graphql"
	"github.com/vektah/gqlparser"
	"github.com/vektah/gqlparser/ast"

	"context"
	"fmt"
	"google.golang.org/grpc"
	"graphql-gateway/generated"
	"graphql-gateway/utils"
	"io"
	"os"
	"time"
)

func getenv(key, fallback string) string {
	value := os.Getenv(key)
	if len(value) == 0 {
		return fallback
	}
	return value
}

var (
	address = getenv("SCHEMA_REGISTRY_URL", "graphql-gateway.registry:81")
)

type schemaResult struct {
	schema *graphql.Schema
	err    error
}

func subscribeToSchema(schemas chan schemaResult) (err error) {
	defer utils.Recovery(&err)

	conn, err := grpc.Dial(address, grpc.WithInsecure())
	if err != nil {
		fmt.Println("error initiating GRPC channel")
		return err
	}
	defer conn.Close()

	gqlSchemaClient := gqlschema.NewGqlSchemaClient(conn)

	stream, err := subscribe(gqlSchemaClient)
	if err != nil {
		fmt.Println("error subscribing to registry", err)
		return err
	}

	for {
		gqlSchemaMessage, err := stream.Recv()

		if err == io.EOF {
			fmt.Println("got EOF")
			break
		}
		if err != nil {
			fmt.Println("error receiving message")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		astSchema, err := parseSdl(source{
			name: "schema registry sdl",
			sdl:  gqlSchemaMessage.Schema,
		})
		if err != nil {
			fmt.Println("error parsing SDL")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		schema, err := ConvertSchema(astSchema)
		if err != nil {
			fmt.Println("error converting schema")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		schemas <- schemaResult{
			schema: schema,
			err:    nil,
		}
	}
	return
}

func subscribe(client gqlschema.GqlSchemaClient) (stream gqlschema.GqlSchema_SubscribeClient, err error) {
	for i := 0; i < 3; i++ {
		stream, err = client.Subscribe(context.Background(), &gqlschema.GqlSchemaSubscribeParams{})
		if err == nil {
			break
		}
		if i+1 < 3 {
			time.Sleep(3000 * time.Millisecond)
		}
	}
	return
}

type source struct {
	name string
	sdl  string
}

func parseSdl(s source) (*ast.Schema, error) {
	schema, err := gqlparser.LoadSchema(&ast.Source{
		Name:  s.name,
		Input: s.sdl,
	})

	if err != nil {
		return nil, err
	}

	return schema, nil
}
