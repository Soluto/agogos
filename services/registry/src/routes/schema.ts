import * as express from "express";
import * as bodyParser from "body-parser";
import sources, { defaultSource } from "../sources-config";
import syncSchema$ from "../sync/sync-service";
import { take } from "rxjs/operators";

const app = express();

const getFromSource = async (source: string, kind: string, res: express.Response) => {
    try {
        const gqlObjects = await sources[source].getGqlObjects(kind);
        res.send(gqlObjects);
    } catch (error) {
        console.warn(`Failed to get ${kind} from source - ${source}`, {
            error
        });
        res.sendStatus(500);
    }
};

const postSource = async (
    source: string,
    kind: string,
    name: string,
    definition: string,
    res: express.Response
) => {
    try {
        await sources[source].registerGqlObject(name, kind, definition);
        res.sendStatus(200);
    } catch (error) {
        console.warn(`Failed to register schema to source - ${source}`, {
            name,
            kind,
            error
        });
        res.sendStatus(500);
    }
};

app
    .use(bodyParser.text())
    .get("/:sourceName/:kind", (req, res) => getFromSource(req.params.sourceName, req.params.kind, res))
    .post("/:sourceName/:kind/:name", (req, res) => {
        const { sourceName, kind, name } = req.params;
        const definition = req.body;
        return postSource(sourceName, kind, name, definition, res);
    })
    .get("/", async (_, res) => res.send(await syncSchema$.pipe(take(1)).toPromise()))
    .post(":kind/:name", (req, res) => {
        const { kind, name } = req.params;
        const definition = req.body;
        return postSource(defaultSource, kind, name, definition, res);
    });

export default app;
