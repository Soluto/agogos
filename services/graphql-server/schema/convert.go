package schema

import (
	"agogos/directives/common"
	"agogos/directives/common/exports"
	"agogos/directives/middlewares"
	"agogos/scalars"
	"agogos/server"
	"agogos/utils"
	"fmt"
	"strings"

	graphql "github.com/graphql-go/graphql"
	"github.com/vektah/gqlparser/ast"

	log "github.com/sirupsen/logrus"
)

type schemaContext struct {
	server.ServerContext
	schema     *ast.Schema
	objects    map[string]*graphql.Object
	interfaces map[string]*graphql.Interface
	unions     map[string]*graphql.Union
	enums      map[string]*graphql.Enum
	inputs     map[string]*graphql.InputObject
}

var errMissingResolver = fmt.Errorf("missing resolver")

func convertType(t *ast.Type, c schemaContext) graphql.Type {
	if t.Elem != nil {
		if t.NonNull {
			return graphql.NewNonNull(graphql.NewList(convertType(t.Elem, c)))
		}
		return graphql.NewList(convertType(t.Elem, c))
	} else if t.NonNull {
		return graphql.NewNonNull(convertNamedType(t, c))
	}

	return convertNamedType(t, c)
}

func convertNamedType(t *ast.Type, c schemaContext) graphql.Type {
	switch name := t.Name(); name {
	case "ID":
		return graphql.ID
	case "String":
		return graphql.String
	case "Int":
		return graphql.Int
	case "Float":
		return graphql.Float
	case "Boolean":
		return graphql.Boolean
	case "Date":
		return graphql.DateTime
	case "JSON":
		return scalars.JSON
	default:
		definition := c.schema.Types[name]

		if definition.Kind == ast.Object {
			return convertSchemaObject(definition, c)
		}
		if definition.Kind == ast.InputObject {
			return convertSchemaInputObject(definition, c)
		}
		if definition.Kind == ast.Interface {
			return convertSchemaInterface(definition, c)
		}
		if definition.Kind == ast.Union {
			return convertSchemaUnion(definition, c)
		}
		if definition.Kind == ast.Scalar {
			log.Panic("Custom scalar handling not implemented")
		}
		if definition.Kind == ast.Enum {
			return convertSchemaEnum(definition, c)
		}

		log.Panic("Unreachable code")
		return nil
	}
}

func convertFieldArgs(a ast.ArgumentDefinitionList, c schemaContext) (graphql.FieldConfigArgument, error) {
	args := make(map[string]*graphql.ArgumentConfig)

	for _, d := range a {
		defaultValue, err := d.DefaultValue.Value(make(map[string]interface{}))

		if err != nil {
			return nil, err
		}

		args[d.Name] = &graphql.ArgumentConfig{
			DefaultValue: defaultValue,
			Description:  d.Description,
			Type:         convertType(d.Type, c),
		}
	}

	return args, nil
}

func createResolver(f *ast.FieldDefinition, c schemaContext) middlewares.Resolver {
	resolver := fieldIdentityResolver

	for _, d := range f.Directives {
		middlewareDefinition, ok := common.MiddlewareDefinitions[d.Name]

		if !ok {
			continue
		}

		middleware := middlewareDefinition.CreateMiddleware(c, f, d)

		resolver = middleware.Wrap(resolver)
	}

	return exports.WrapperMiddleware.Wrap(resolver)
}

func fieldIdentityResolver(rp graphql.ResolveParams) (interface{}, error) {
	fieldName := rp.Info.FieldName
	return utils.IdentityResolver(fieldName, rp.Source)
}

func convertSchemaField(f *ast.FieldDefinition, c schemaContext) (*graphql.Field, error) {
	args, err := convertFieldArgs(f.Arguments, c)

	if err != nil {
		return nil, err
	}

	return &graphql.Field{
		Description: f.Description,
		Type:        convertType(f.Type, c),
		Resolve:     createResolver(f, c),
		Args:        args,
	}, nil
}

func convertSchemaEnum(d *ast.Definition, c schemaContext) *graphql.Enum {
	enum, ok := c.enums[d.Name]

	if ok {
		return enum
	}

	values := make(map[string]*graphql.EnumValueConfig)

	for _, v := range d.EnumValues {
		enumValue := new(graphql.EnumValueConfig)
		enumValue.Value = v.Name
		enumValue.Description = v.Description
		values[v.Name] = enumValue
	}

	enum = graphql.NewEnum(graphql.EnumConfig{Name: d.Name, Values: values, Description: d.Description})

	c.enums[d.Name] = enum

	return enum
}

func convertSchemaInputObject(d *ast.Definition, c schemaContext) *graphql.InputObject {
	object, ok := c.inputs[d.Name]

	if ok {
		return object
	}

	fieldsThunk := graphql.InputObjectConfigFieldMapThunk(func() graphql.InputObjectConfigFieldMap {
		fields := make(map[string]*graphql.InputObjectFieldConfig)
		for _, field := range d.Fields {
			if field.Name[:2] == "__" {
				continue
			}

			fields[field.Name] = convertSchemaInputField(field, c)
		}
		return fields
	})

	object = graphql.NewInputObject(graphql.InputObjectConfig{Name: d.Name, Fields: fieldsThunk, Description: d.Description})

	c.inputs[d.Name] = object

	return object
}

func convertSchemaInputField(f *ast.FieldDefinition, c schemaContext) *graphql.InputObjectFieldConfig {
	return &graphql.InputObjectFieldConfig{
		Description: f.Description,
		Type:        convertType(f.Type, c),
	}
}

func convertSchemaUnion(d *ast.Definition, c schemaContext) *graphql.Union {
	union, ok := c.unions[d.Name]

	if ok {
		return union
	}

	types := []*graphql.Object{}

	for _, t := range d.Types {
		types = append(types, convertSchemaObject(c.schema.Types[t], c))
	}

	union = graphql.NewUnion(graphql.UnionConfig{Name: d.Name, Types: types, Description: d.Description})

	c.unions[d.Name] = union

	return union
}

func convertSchemaInterface(d *ast.Definition, c schemaContext) *graphql.Interface {
	object, ok := c.interfaces[d.Name]

	if ok {
		return object
	}

	fieldsThunk := graphql.FieldsThunk(func() graphql.Fields {
		fields := make(map[string]*graphql.Field)
		for _, field := range d.Fields {
			if strings.HasPrefix(field.Name, "__") {
				continue
			}

			schemaField, err := convertSchemaField(field, c)
			if err != nil {
				log.Panic(err)
			}

			fields[field.Name] = schemaField
		}
		return fields
	})

	object = graphql.NewInterface(graphql.InterfaceConfig{Name: d.Name, Fields: fieldsThunk, Description: d.Description})

	c.interfaces[d.Name] = object

	return object
}

func convertSchemaObject(d *ast.Definition, c schemaContext) *graphql.Object {
	if d == nil {
		return nil
	}

	object, ok := c.objects[d.Name]

	if ok {
		return object
	}

	fieldsThunk := graphql.FieldsThunk(func() graphql.Fields {
		fields := make(map[string]*graphql.Field)
		for _, field := range d.Fields {
			if field.Name[:2] == "__" {
				continue
			}

			schemaField, err := convertSchemaField(field, c)
			if err != nil {
				log.Panic(err)
			}

			fields[field.Name] = schemaField
		}
		return fields
	})

	object = graphql.NewObject(graphql.ObjectConfig{Name: d.Name, Fields: fieldsThunk, Description: d.Description})

	c.objects[d.Name] = object

	return object
}

// ConvertSchema converts schema definition to a graphql schema
func convertSchema(serverContext server.ServerContext, astSchema *ast.Schema) (schemaPtr *graphql.Schema, err error) {
	defer utils.Recovery(&err)

	schemaCtx := schemaContext{
		serverContext,
		astSchema,
		make(map[string]*graphql.Object),
		make(map[string]*graphql.Interface),
		make(map[string]*graphql.Union),
		make(map[string]*graphql.Enum),
		make(map[string]*graphql.InputObject),
	}

	schema, err := graphql.NewSchema(graphql.SchemaConfig{
		Query:    convertSchemaObject(astSchema.Query, schemaCtx),
		Mutation: convertSchemaObject(astSchema.Mutation, schemaCtx),
	})

	schemaPtr = new(graphql.Schema)
	*schemaPtr = schema

	return
}
