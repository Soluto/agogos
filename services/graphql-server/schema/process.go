package schema

import (
	"agogos/extensions/upstreams"
	"agogos/extensions/upstreams/authentication"
	agogos "agogos/generated"
	"agogos/server"

	"github.com/graphql-go/graphql"
	log "github.com/sirupsen/logrus"
	"github.com/vektah/gqlparser"
	"github.com/vektah/gqlparser/ast"
)

func parseSdl(name, sdl string) (*ast.Schema, error) {
	schema, err := gqlparser.LoadSchema(&ast.Source{
		Name:  name,
		Input: sdl,
	})

	if err != nil {
		return nil, err
	}

	return schema, nil
}

func transformToSchema(config *agogos.ConfigurationMessage) (*graphql.Schema, error) {
	astSchema, err := parseSdl(
		"schema registry sdl",
		config.Schema.Definition,
	)
	if err != nil {
		log.WithError(err).WithField("schema", config.Schema.Definition).Warn("Error parsing SDL")
		return nil, err
	}

	upstreamsMap := createUpstreams(config.Upstreams, config.UpstreamAuthCredentials)
	serverContext := server.CreateServerContext(upstreamsMap)

	convertedSchema, err := convertSchema(serverContext, astSchema)
	if err != nil {
		log.WithError(err).Warn("Error converting schema")
		return nil, err
	}

	return convertedSchema, nil
}

func createUpstreams(upsConfigs []*agogos.Upstream, upsAuthConfigs []*agogos.UpstreamAuthCredentials) map[string]upstreams.Upstream {
	upsAuths := make(map[string]map[string]authentication.UpstreamAuthentication)
	for _, upsAuthConfig := range upsAuthConfigs {
		if _, ok := upsAuths[upsAuthConfig.AuthType]; !ok {
			upsAuths[upsAuthConfig.AuthType] = make(map[string]authentication.UpstreamAuthentication)
		}

		upsAuths[upsAuthConfig.AuthType][upsAuthConfig.Authority] = authentication.CreateFromConfig(upsAuthConfig)
	}

	ups := make(map[string]upstreams.Upstream, len(upsConfigs))

	for _, upConfig := range upsConfigs {
		var matchedUpsAuth authentication.UpstreamAuthentication
		if m, ok := upsAuths[upConfig.Auth.AuthType]; ok {
			if upsAuth, ok := m[upConfig.Auth.Authority]; ok {
				matchedUpsAuth = upsAuth
			}
		}

		ups[upConfig.Host] = upstreams.CreateFromConfig(upConfig, matchedUpsAuth)
	}

	return ups
}

type SchemaResult struct {
	Schema *graphql.Schema
	Error  error
}

func Process(configCh chan *agogos.ConfigurationMessage) chan SchemaResult {
	schemaCh := make(chan SchemaResult)

	go func() {
		defer close(schemaCh)

		for {
			config := <-configCh
			log.Infoln("Got new configuration from registry")
			schema, err := transformToSchema(config)
			schemaCh <- SchemaResult{
				Schema: schema,
				Error:  err,
			}
		}
	}()

	return schemaCh
}
