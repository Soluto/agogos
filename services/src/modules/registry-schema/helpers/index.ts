import proceedPolicy from './policy-helper';

export { proceedPolicy };
