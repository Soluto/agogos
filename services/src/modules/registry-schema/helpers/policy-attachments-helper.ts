import * as _ from 'lodash';
import * as opaHelper from '../../opa-helper';
import logger from '../../logger';
import { PolicyType, ResourceMetadata, ResourceRepository } from '../../resource-repository';
import resourceRepository from '../repository';
import { ResourceMetadataInput, PolicyInput } from '../types';

type TempLocalPolicyAttachment = { metadata: ResourceMetadataInput; path: string; type: PolicyType };

const policyMetadataComparer = (p1: PolicyInput, p2: PolicyInput) =>
  p1.metadata.namespace === p2.metadata.namespace && p1.metadata.name === p2.metadata.name;

export default class PolicyAttachmentsHelper {
  protected policyAttachmentsToSave: TempLocalPolicyAttachment[] = [];
  protected policyAttachmentsToDelete: PolicyInput[] = [];

  async sync(oldState: PolicyInput[], newState: PolicyInput[]) {
    // Policy definitions generated by plugins will always be re-generated here. This is probably
    // better than reading the previous gateway resource group just for this comparison
    const addedOrUpdatedPolicies = _.differenceWith(newState, oldState, _.isEqual);
    await this.generate(addedOrUpdatedPolicies);

    const removedPolicies = _.differenceWith(oldState, newState, policyMetadataComparer);
    this.policyAttachmentsToDelete.push(...removedPolicies);
  }

  private async generate(policies: PolicyInput[]) {
    for (const policy of policies) {
      if (!policyAttachmentStrategies[policy.type]) continue;

      const attachment = await policyAttachmentStrategies[policy.type].generate(policy);
      this.policyAttachmentsToSave.push(attachment);
    }
  }

  async writeToRepo() {
    for (const attachment of this.policyAttachmentsToSave) {
      await policyAttachmentStrategies[attachment.type].saveToRepo(resourceRepository, attachment);
    }

    for (const policy of this.policyAttachmentsToDelete) {
      await policyAttachmentStrategies[policy.type].removeFromRepo(resourceRepository, policy.metadata);
    }
  }

  async cleanup() {
    for (const attachment of this.policyAttachmentsToSave) {
      await policyAttachmentStrategies[attachment.type].cleanup(attachment);
    }
  }
}

const policyAttachmentStrategies = {
  [PolicyType.opa]: {
    generate: async (input: PolicyInput) => {
      const path = await opaHelper.prepareCompiledRegoFile(input.metadata, input.code);
      return { path, metadata: input.metadata, type: PolicyType.opa };
    },
    cleanup: async (attachment: TempLocalPolicyAttachment) => {
      try {
        await opaHelper.deleteLocalRegoFile(attachment.path);
      } catch (err) {
        logger.warn(
          { err, attachment },
          'Failed cleanup of compiled rego file, this did not affect the request outcome'
        );
      }
    },
    saveToRepo: async (resourceRepository: ResourceRepository, attachment: TempLocalPolicyAttachment) => {
      const compiledRego = await opaHelper.readLocalRegoFile(attachment.path);
      const filename = opaHelper.getCompiledFilename(attachment.metadata);
      await resourceRepository.writePolicyAttachment(filename, compiledRego);
    },
    removeFromRepo: async (resourceRepository: ResourceRepository, metadata: ResourceMetadata) => {
      const filename = opaHelper.getCompiledFilename(metadata);
      await resourceRepository.deletePolicyAttachment(filename);
    },
  },
};
