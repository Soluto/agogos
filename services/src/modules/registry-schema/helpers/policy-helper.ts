import { parseType } from 'graphql';
import { PolicyInput } from '..';
import logger from '../../logger';
import { PolicyArgDefinition } from '../../resource-repository';

export function markOptionalPolicyArgs(policies?: PolicyInput[]) {
  policies?.forEach(p => {
    if (!p.args) return;

    Object.entries(p.args).forEach(([argName, argDef]: [string, PolicyArgDefinition]) => {
      try {
        const type = parseType(argDef.type);
        argDef.optional = type.kind !== 'NonNullType' || !!argDef.default;
      } catch (err) {
        logger.warn({ policy: p.metadata, argName, argDef, err }, 'Invalid policy argument');
        throw err;
      }
    });
  });
}
