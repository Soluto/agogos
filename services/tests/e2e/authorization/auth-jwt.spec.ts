import { GraphQLClient } from 'graphql-request';
import { updateGatewaySchema } from '../../helpers/utility';
import {
  RegistryMutationResponse,
  updatePoliciesMutation,
  updateSchemasMutation,
} from '../../helpers/registry-request-builder';
import GraphQLErrorSerializer from '../../utils/graphql-error-serializer';
import { getToken } from '../../helpers/get-token';
import { schema, policies, getUserQuery, arbitraryDataQuery } from './auth-jwt.schema';

const gatewayClient = new GraphQLClient('http://localhost:8080/graphql');
const registryClient = new GraphQLClient('http://localhost:8090/graphql');

describe('authorization', () => {
  let defaultAccessToken: string;

  beforeAll(async () => {
    expect.addSnapshotSerializer(GraphQLErrorSerializer);

    const registryAccessToken = await getToken({ scope: 'stitch-registry' });
    registryClient.setHeader('Authorization', `Bearer ${registryAccessToken}`);

    defaultAccessToken = await getToken();
  });

  beforeEach(() => {
    gatewayClient.setHeader('Authorization', `Bearer ${defaultAccessToken}`);
  });

  // This is kind of both the "before" section and a test, but it was weird putting a test in an actual before section
  it('creates the policy and schema resources', async () => {
    const policiesResponse = await registryClient.request<RegistryMutationResponse>(updatePoliciesMutation, {
      policies,
    });
    expect(policiesResponse.result.success).toBeTruthy();

    const schemaResponse = await registryClient.request<RegistryMutationResponse>(updateSchemasMutation, { schema });
    expect(schemaResponse.result.success).toBeTruthy();

    const resp = await updateGatewaySchema('http://localhost:8080');
    expect(resp.status).toEqual(200);
  });

  it('allows access to a field based on an argument using param injection from source', async () => {
    const response = await gatewayClient.request(getUserQuery('userAdmin')).catch(err => err.response);
    expect(response).toMatchSnapshot();
  });

  it('rejects access to a field when policy test fails, but still returns the other fields', async () => {
    const response = await gatewayClient.request(getUserQuery('user')).catch(err => err.response);
    expect(response).toMatchSnapshot();
  });

  it('allows access to a field based on a non-default argument', async () => {
    const response = await gatewayClient.request(getUserQuery('userIgnoreRole')).catch(err => err.response);
    expect(response).toMatchSnapshot();
  });

  it('rejects access to a field based on JWT info when no JWT is sent', async () => {
    const response = await gatewayClient.request(arbitraryDataQuery).catch(err => err.response);
    expect(response).toMatchSnapshot();
  });

  it('rejects access to a field based on JWT info', async () => {
    const accessToken = await getToken({ clientId: 'e2e-jwt-disallowed-client-id' });
    gatewayClient.setHeader('Authorization', `Bearer ${accessToken}`);
    const response = await gatewayClient.request(arbitraryDataQuery).catch(err => err.response);
    expect(response).toMatchSnapshot();
  });

  it('allows access to a field based on JWT info', async () => {
    const accessToken = await getToken({ clientId: 'e2e-jwt-allowed-client-id' });
    gatewayClient.setHeader('Authorization', `Bearer ${accessToken}`);
    const response: any = await gatewayClient.request(arbitraryDataQuery);
    expect(response).toMatchSnapshot();
  });
});
