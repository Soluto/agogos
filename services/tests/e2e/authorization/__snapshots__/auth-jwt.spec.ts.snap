// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authorization allows access to a field based on JWT info 1`] = `
Object {
  "arbitraryData": Object {
    "arbitraryField": "arbitraryValue",
  },
}
`;

exports[`authorization allows access to a field based on a non-default argument 1`] = `
Object {
  "userIgnoreRole": Object {
    "firstName": "John",
    "lastName": "Smith",
    "role": "normal",
  },
}
`;

exports[`authorization allows access to a field based on an argument using param injection from source 1`] = `
Object {
  "userAdmin": Object {
    "firstName": "John",
    "lastName": "Smith",
    "role": "admin",
  },
}
`;

exports[`authorization rejects access to a field based on JWT info 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy jwtName in namespace auth_jwt",
      "errorCode": "UNAUTHORIZED_BY_POLICY",
      "path": [
        "arbitraryData",
        "arbitraryField"
      ]
    }
  ],
  "data": {
    "arbitraryData": {
      "arbitraryField": null
    }
  }
}
`;

exports[`authorization rejects access to a field based on JWT info when no JWT is sent 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy jwtName in namespace auth_jwt",
      "errorCode": "UNAUTHORIZED_BY_POLICY",
      "path": [
        "arbitraryData",
        "arbitraryField"
      ]
    }
  ],
  "data": {
    "arbitraryData": {
      "arbitraryField": null
    }
  }
}
`;

exports[`authorization rejects access to a field when policy test fails, but still returns the other fields 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy onlyAdmin in namespace auth_jwt",
      "errorCode": "UNAUTHORIZED_BY_POLICY",
      "path": [
        "user",
        "lastName"
      ]
    }
  ],
  "data": {
    "user": {
      "firstName": "John",
      "lastName": null,
      "role": "normal"
    }
  }
}
`;
